cai_causal_graph/utils.py

```diff
@@ -0,0 +1,112 @@
+        'X future(n=2)' -> 'X', 2 if lagged in the future.
+
+    :param node_name: The name of the node.
+    :return: A tuple with the variable name and lag.
+    """
+    # get the string name if the node is a Node object
+    if isinstance(node_name, HasIdentifier):
+        node_name = node_name.identifier
+
+    if not isinstance(node_name, str):
+        raise TypeError(f'Expected node name to be a string, got type {type(node_name)}.')
+
+    # Validation for cases with both "lag" and "future"
+    if 'lag' in node_name and 'future' in node_name:
+        raise ValueError(f'Invalid node name with both past and future lags: {node_name}.')
+
+    is_match = re.match(r'^(\w+)(?: lag\(n=(\d+)\))?(?: future\(n=(\d+)\))?$', node_name)
+
+    if is_match:
+        variable_name = is_match.group(1)
+        past_lag = is_match.group(2)
+        future_lag = is_match.group(3)
+
+        # # we need to check that the node name does not contain both a lag and future lag
+        if past_lag:
+            return variable_name, -int(past_lag)
+        elif future_lag:
+            return variable_name, int(future_lag)
+        else:
+            return variable_name, 0  # no lag information
+
+    else:
+        raise ValueError(f'Invalid node name: {node_name}. Expected format: "X", "X lag(n=2)", or "X future(n=2)".')
+
+
+def get_name_with_lag(variable_or_node_name: str, lag: int) -> str:
+    """
+    Get the name of a lagged node.
+
+    If the lag is 0, then the variable name is returned.
+    If the lag is not 0, then the old lag is removed and
+    variable name is appended with the lag.
+
+    Example:
+        'X', -2 -> 'X lag(n=2)', # lagged in the past
+        'X', 2 -> 'X future(n=2)', # lagged in the future
+
+    :param variable_or_node_name: The name of the variable or node.
+    :param lag: The lag of the node.
+    :return: The name of the lagged node.
+    """
+    # remove the old lag if it exists
+    variable_name, old_lag = get_variable_name_and_lag(variable_or_node_name)
+
+    assert isinstance(lag, int), f'Expected lag to be an integer, got type {type(lag)}.'
+    if lag == 0:
+        return variable_name
+    elif lag > 0:
+        return f'{variable_name} future(n={lag})'
+    else:
+        return f'{variable_name} lag(n={-lag})'
+
+
+def extract_names_and_lags(
+    node_names: List[NodeLike],
+) -> Tuple[List[Dict[str, int]], int]:
+    """
+    Extract the names and lags from a list of node names.
+    This is useful for converting a list of node names into a list of variable names and lags.
+
+    Example:
+        ['X', 'Y', 'Z lag(n=2)'] -> [{'X': 0}, {'Y': 0}, {'Z': 2}]
+
+    :param node_names: List of node names.
+    :return: List of dictionaries with variable names and lags and the maximum lag.
+    """
+    names_and_lags = []
+    max_lag: int = 0
+    for node_name in node_names:
+        variable_name, lag = get_variable_name_and_lag(node_name)
+        if abs(lag) > abs(max_lag):
```

Comment:
Why is this absolute value? Should we not look at positive and negative lags and potentially return both?

Reply:
Added JIRA ticket to double check this in future PR.